File: dataset/random/src/problems/multi_knapsack.cpp
#include <vector>
#include <string>

#include "utils.h"

Problem generate_problem(const int id, const std::pair<int, int> &items, const std::pair<int, int> &knapsacks,
                         const std::pair<double, double> &knapsack_capacity,
                         const std::pair<double, double> &item_profit,
                         const std::pair<double, double> &item_weight) {
    const int n_knapsacks = generate_random_int(knapsacks.first, knapsacks.second);
    const int n_items = generate_random_int(items.first, items.second);
    std::vector<double> knapsack_capacities(n_knapsacks);
    for (int j = 0; j < n_knapsacks; ++j) {
        knapsack_capacities[j] = generate_random_double(knapsack_capacity.first, knapsack_capacity.second);
    }
    std::vector<double> item_profits(n_items);
    for (int i = 0; i < n_items; ++i) {
        item_profits[i] = generate_random_double(item_profit.first, item_profit.second);
    }
    std::vector<double> item_weights(n_items);
    for (int i = 0; i < n_items; ++i) {
        item_weights[i] = generate_random_double(item_weight.first, item_weight.second);
    }

    const int n_vars = n_items * n_knapsacks; // x_ij
    std::vector<double> c(n_vars);
    for (int i = 0; i < n_items; ++i) {
        for (int j = 0; j < n_knapsacks; ++j) {
            // convert this to minimization problem
            c[i * n_knapsacks + j] = -item_profits[i];
        }
    }

    std::vector<std::vector<double> > A;
    std::vector<double> b;
    std::vector<char> types;

    // Knapsack capacity constraints: sum_{i} w_ij * x_ij <= W_j for all j
    for (int j = 0; j < n_knapsacks; ++j) {
        std::vector<double> row(n_vars, 0.0);
        for (int i = 0; i < n_items; ++i) {
            row[i * n_knapsacks + j] = item_weights[i];
        }
        A.push_back(row);
        b.push_back(knapsack_capacities[j]);
        types.push_back('L');
    }

    // Item selection constraints: sum_{j} x_ij = 1 for all i
    for (int i = 0; i < n_items; ++i) {
        std::vector<double> row(n_vars, 0.0);
        for (int j = 0; j < n_knapsacks; ++j) {
            row[i * n_knapsacks + j] = 1.0;
        }
        A.push_back(row);
        b.push_back(1.0);
        types.push_back('E');
    }

    std::vector<double> lb(n_vars, 0.0);
    std::vector<double> ub(n_vars, 1.0);

    Problem problem = {"random_MKP_" + std::to_string(id), types, c, A, b, lb, ub};
    solve(problem);
    return problem;
}

std::vector<Problem> multi_knapsack(const int n_problems, const std::pair<int, int> &items,
                                    const std::pair<int, int> &knapsacks,
                                    const std::pair<double, double> &knapsack_capacity,
                                    const std::pair<double, double> &item_profit,
                                    const std::pair<double, double> &item_weight) {
    std::vector<Problem> problems;
    for (int i = 0; i < n_problems; ++i) {
        problems.push_back(generate_problem(i, items, knapsacks, knapsack_capacity, item_profit, item_weight));
    }
    return problems;
}

File: dataset/random_probs/multi_knapsack.py
import numpy as np

"""
Formulation
max sum_i^n sum_{j in N_i} p_ij * x_ij
s.t.
    sum_i^k sum_{j in N_i} w_ij * x_ij <= W_j
    sum_{j in N_i} x_ij = 1 forall i
    x_ij in {0, 1} forall j in N_i, forall i
"""


def multi_knapsack(n_problems: int, items: (int, int), knapsacks: (int, int), knapsack_capacity: (float, float),
                   item_profit: (float, float), item_weight: (float, float)):
    return [__generate_problem(i, items, knapsacks, knapsack_capacity, item_profit, item_weight) for
            i in range(n_problems)]


def __generate_problem(id: int, items: (int, int), knapsacks: (int, int), knapsack_capacity: (float, float),
                   item_profit: (float, float), item_weight: (float, float)):
    n_knapsacks = np.random.randint(*knapsacks)
    n_items = np.random.randint(*items)
    knapsack_capacities = np.random.uniform(*knapsack_capacity, n_knapsacks)
    item_profits = np.random.uniform(*item_profit, n_items)
    item_weights = np.random.uniform(*item_weight, n_items)

    n_vars = n_items * n_knapsacks  # x_ij
    c = -np.tile(item_profits, n_knapsacks)  # convert to minimiz. problem

    A, b, types = [], [], []

    ## constraints
    # sum_i^k sum_{j in N_i} w_ij * x_ij <= W_j
    for j in range(n_knapsacks):
        row = np.zeros(n_vars)
        for i in range(n_items):
            row[i * n_knapsacks + j] = item_weights[i]
        A.append(row)
        b.append(knapsack_capacities[j])
        types.append('L')

    # sum_{j in N_i} x_ij <= 1 forall i
    for i in range(n_items):
        row = np.zeros(n_vars)
        for j in range(n_knapsacks):
            row[i * n_knapsacks + j] = 1
        A.append(row)
        b.append(1)
        types.append('E')

    A = np.array(A)
    b = np.array(b)
    types = np.array(types)
    bnd = [{"LO": 0, "UP": 1} for _ in range(n_vars)]

    return f"random_MKP_{id}", types, c, A, b, bnd
